bridgenode/rev.go

```diff
@@ -62,19 +64,166 @@ func BlockAndRevReader(
+
+	var datFileNum uint32
+	// minOffset holds the smallest block offset that was seen
+	minOffset := uint32(math.MaxUint32)
+	// maxOffset holds the largest block offset that was seen
+	maxOffset := uint32(0)
+	// minRevOffset holds the smallest rev offset that was seen
+	minRevOffset := uint32(math.MaxUint32)
+	// maxRevOffset holds the largest rev offset that was seen
+	maxRevOffset := uint32(0)
+	// offsetsRead holds the number of blocks < count that are loacated in
+	// blk file with number `datFileNum`
+	offsetsRead := uint32(0)
+	for ; offsetsRead < uint32(count); offsetsRead++ {
+		var datFileNumTmp uint32
+		err = binary.Read(offsetReader, binary.BigEndian, &datFileNumTmp)
+		if err != nil {
+			break
+		}
+		if offsetsRead > 0 && datFileNumTmp != datFileNum {
+			// break if the block is located in a different blk file
+			break
+		}
+		datFileNum = datFileNumTmp
+		err = binary.Read(offsetReader, binary.BigEndian, &offsets[offsetsRead])
+		if err != nil {
+			return
+		}
+		err = binary.Read(offsetReader, binary.BigEndian, &revOffsets[offsetsRead])
+		if err != nil {
+			return
+		}
+
+		if offsets[offsetsRead] < minOffset {
+			minOffset = offsets[offsetsRead]
+		}
+		if offsets[offsetsRead] > maxOffset {
+			maxOffset = offsets[offsetsRead]
+		}
+
+		if revOffsets[offsetsRead] < minRevOffset {
+			minRevOffset = revOffsets[offsetsRead]
+		}
+		if revOffsets[offsetsRead] > maxRevOffset {
+			maxRevOffset = revOffsets[offsetsRead]
+		}
+	}
+
+	if offsetsRead == 0 {
+		return
+	}
+
+	blockFile, err := os.Open(filepath.Join(blockDir,
+		fmt.Sprintf("blk%05d.dat", datFileNum)))
+	if err != nil {
+		return
+	}
+	defer blockFile.Close()
+	blockFileInfo, _ := blockFile.Stat()
+
+	// Read all block data needed for the blocks into memory.
+	// The blocks that are needed are located between the minOffset and
+	// maxOffset+(the size of the block at maxOffset).
+	// 1<<23 = +8MB TODO: whats the max on disk size of a block?
+	blockData := make([]byte, min(maxOffset+1<<23, uint32(blockFileInfo.Size()))-minOffset)
+	_, err = blockFile.ReadAt(blockData, int64(minOffset))
+	if err != nil {
+		return
+	}
+
+	revFile, err := os.Open(filepath.Join(blockDir,
+		fmt.Sprintf("rev%05d.dat", datFileNum)))
+	if err != nil {
+		return
+	}
+	defer revFile.Close()
+	revFileInfo, _ := revFile.Stat()
+
+	// Read all rev data needed for the blocks into memory.
+	// The rev data thats needed is located between the minRevOffset and
+	// maxRevOffset+(the size of the rev at maxRevOffset).
+	// 1<<23 = +8MB TODO: whats the max on disk size of a rev?
+	revData := make([]byte, min(maxRevOffset+1<<23, uint32(revFileInfo.Size()))-minRevOffset)
+	_, err = revFile.ReadAt(revData, int64(minRevOffset))
+	if err != nil {
+		return
+	}
+
+	blocks = make([]wire.MsgBlock, offsetsRead)
+	revs = make([]RevBlock, offsetsRead)
+	skip := make([]byte, 8)
+	for i := uint32(0); i < offsetsRead; i++ {
+		blockBuf := bytes.NewBuffer(blockData[offsets[i]-minOffset:])
```

Comment:
These aren't really ideal since a new buffer per loop does put some gc pressure. It's small now as the other bottlenecks take so much time so it's fine for now but something to note for the future.

Reply:
i am only using the buffer to have a io.Reader to pass to `MsgBlock.Deserialize`. bytes.Buffer sadly does not support Seek otherwise one buffer would be enough.